src/ompl/base/samplers/InformedStateSampler.h

```diff
@@ -0,0 +1,109 @@
+
+/* Authors: Jonathan Gammell */
+
+#ifndef OMPL_BASE_SAMPLERS_INFORMED_SAMPLER_
+#define OMPL_BASE_SAMPLERS_INFORMED_SAMPLER_
+
+// We inherit from StateSampler
+#include "ompl/base/StateSampler.h"
+// Deriving functions must be able to sample within a given cost
+#include "ompl/base/Cost.h"
+// We use a pointer to the problem definition to access problem and solution data.
+#include "ompl/base/ProblemDefinition.h"
+
+namespace ompl
+{
+    namespace base
+    {
+        OMPL_CLASS_FORWARD(InformedStateSampler);
+
+        /** \brief An abstract class for state samplers that use information
+        about the current solution to limit future search to a planning
+        subproblem that contains all possibly better solutions. */
+        class InformedStateSampler : public StateSampler
+        {
+        public:
+            /** \brief Construct a sampler that only generates states with a heuristic solution estimate that is less than the cost of the current solution. */
+            InformedStateSampler(const StateSpace* space, const ProblemDefinitionPtr probDefn, const Cost* bestCost);
+            virtual ~InformedStateSampler(void)
+            {
+            }
+
+            /** \brief Sample uniformly in the subset of the state space whose heuristic solution estimates are less than the current best cost (as defined by the pointer passed at construction). By default just calls sampleUniform(State*, Cost) with cost given by the member variable. */
+            virtual void sampleUniform(State* statePtr);
+
+            /** \brief Sample uniformly in the subset of the state space whose heuristic solution estimates are less than the provided cost. */
+            virtual void sampleUniform(State* statePtr, const Cost& maxCost) = 0;
+
+            /** \brief Sample uniformly in the subset of the state space whose heuristic solution estimates are between the provided costs. */
+            virtual void sampleUniform(State* statePtr, const Cost& minCost, const Cost& maxCost) = 0;
+
+            /** \brief Whether the sampler can provide a measure of the informed subset */
+            virtual bool hasInformedMeasure() const = 0;
+
+            /** \brief The measure of the subset of the state space defined by the current solution cost that is being searched. Does not consider problem boundaries but returns the measure of the entire space if no solution has been found or if a closed form expression for the measure does not exist. By default calls the 1-argument overloaded version with the current best cost; however, there may be more efficient ways to do this for some cost functions. */
+            virtual double getInformedMeasure() const;
+
+            /** \brief The measure of the subset of the state space defined by the current solution cost that is being searched. Does not consider problem boundaries but returns the measure of the entire space if no solution has been found or if a closed form expression for the measure does not exist. */
```

Comment:
Is this measure specific to the sampling strategy? I am not sure the measure functions belong here. It almost seems like this should be a new concept / new object in OMPL. Or perhaps adding it to OptimizationObjective? we should discuss.